### PR TITLE
Support Npu graph executor

### PR DESCRIPTION
**Pull request by joozhan**
## Motivation
Support Npu Graph Runner with torch.compile and Npu Graph Engine

## Modifications
1. Abstract the DeviceRunnerBase class for interfacing with different hardware, extracting common abstract methods such as initialize, prepare_forward_batch, warm_up, can_run_graph, and replay.

2. The CudaGraphRunner subclass inherits from DeviceRunnerBase and implements CUDA device-related abstract methods.

3. Provide a foundational subclass NpuGraphRunner that supports interfacing the compile graph engine with eager mode.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/model_executor/cuda_graph_runner.py [hunk 1]
```diff
@@ -365,7 +532,7 @@ def can_run(self, forward_batch: ForwardBatch):
         is_tbo_supported = (
             forward_batch.can_run_tbo if self.enable_two_batch_overlap else True
         )
-
+        print("forward_batch.batch_size：", forward_batch.batch_size)
```
Comment (zhyncs) [inline python/sglang/srt/model_executor/cuda_graph_runner.py]:
```suggestion
```

File: python/sglang/srt/model_executor/npu_graph_runner.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+                run_once()
+
+        return
+
+    @contextmanager
+    def compile_context(self, forward_batch: ForwardBatch):
+        if self.enable_torch_compile and forward_batch.batch_size in self.compile_bs:
+            yield self.model_runner.model.compile_forward
+        else:
+            yield self.model_runner.model.forward
```
Comment (zhyncs) [inline python/sglang/srt/model_executor/npu_graph_runner.py]:
```suggestion
            yield self.model_runner.model.forward
            
```

File: python/sglang/srt/model_executor/npu_graph_runner.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
```
Comment (zhyncs) [inline python/sglang/srt/model_executor/npu_graph_runner.py]:
```suggestion
```

Comment (joozhan) [reply python/sglang/srt/model_executor/npu_graph_runner.py]: Hi, I don't understand the meaning of this review comment. Currently, in NPU graph mode, I've adopted a batch size binning approach `torchair.inference.set_dim_gears(forward_batch.input_ids, dim_gears={0: self.compile_bs}) ` to support graph integration. Specifically:
```
self.model_runner.model.compile_forward = torch.compile(
    torch.no_grad()(self.model_runner.model.forward),
    fullgraph=True,
    backend=get_compiler_backend(),
    mode=compile_mode
)
```
The current logic implies: if the batch size during execution is supported by the compilation in the forward pass, it will use graph mode; otherwise, it will fall back to eager mode. Could you please provide more detailed feedback? Thank you.

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 4]
```diff
@@ -36,12 +36,20 @@
...
 logger = logging.getLogger(__name__)
 
 GB = 1024 * 1024 * 1024
 _is_cuda = is_cuda()
+_is_npu = is_npu()
+
+if is_npu():
+    try:
+        import torch_npu
+    except ImportError:
```
Comment (Alcanderian) [inline python/sglang/srt/mem_cache/memory_pool.py]: remove try as last PR

File: python/sglang/srt/model_executor/cuda_graph_runner.py
Comment (Alcanderian) [inline python/sglang/srt/model_executor/cuda_graph_runner.py:1]: please fix code style with pre-commit linter

File: python/sglang/srt/model_executor/cuda_graph_runner.py [hunk 5]
```diff
@@ -389,6 +551,7 @@ def can_run(self, forward_batch: ForwardBatch):
             and is_encoder_lens_supported
             and is_tbo_supported
             and capture_hidden_mode_matches
+            and forward_batch.forward_mode.is_cuda_graph()
```
Comment (Alcanderian) [inline python/sglang/srt/model_executor/cuda_graph_runner.py]: Could we move this line to the first line of this method? To skip other logics

File: python/sglang/srt/utils.py
Comment (Alcanderian) [inline python/sglang/srt/utils.py:1]: lint

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 6]
```diff
@@ -487,8 +488,14 @@ def set_kv_buffer(
                 self.v_buffer[layer_id - self.start_layer][loc] = cache_v
             current_stream.wait_stream(self.alt_stream)
         else:
-            self.k_buffer[layer_id - self.start_layer][loc] = cache_k
-            self.v_buffer[layer_id - self.start_layer][loc] = cache_v
+            if _is_npu:
+                import torch_npu
+                assert loc.size() > 0, f"expect loc.size() > 0, but got {loc.size()}"
+                torch_npu.scatter_update_(self.k_buffer[layer_id - self.start_layer].unsqueeze(0), loc[0], cache_k.unsqueeze(0))
+                torch_npu.scatter_update_(self.v_buffer[layer_id - self.start_layer].unsqueeze(0), loc[0], cache_v.unsqueeze(0))
```
[the remark this answers is not in the document]
Comment (joozhan) [reply python/sglang/srt/mem_cache/memory_pool.py]: yes

File: python/sglang/srt/utils.py
Comment (joozhan) [reply python/sglang/srt/utils.py:1]: done

File: python/sglang/srt/model_executor/cuda_graph_runner.py (hunk 5 above)
Comment (joozhan) [reply python/sglang/srt/model_executor/cuda_graph_runner.py]: ok, done

File: python/sglang/srt/model_executor/cuda_graph_runner.py
Comment (joozhan) [reply python/sglang/srt/model_executor/cuda_graph_runner.py:1]: done

File: python/sglang/srt/mem_cache/memory_pool.py (hunk 4 above)
Comment (joozhan) [reply python/sglang/srt/mem_cache/memory_pool.py]: done

File: python/sglang/srt/utils.py [hunk 7]
```diff
@@ -1525,14 +1525,15 @@ def get_device_capability(device_id: int = 0) -> Tuple[int, int]:
...
-    }
-    return config
+npu_compile_config = {}
+def set_npu_compiler_config(compile_config: Dict[str, Any] = None):
+    global npu_compile_config
+    for key, value in compile_config.items():
+        if key in npu_compile_config and value != npu_compile_config[key]:
+            raise ValueError(
+                f"Cannot set {key} to {value} in npu_compile_config, it is already set to {npu_compile_config[key]}."
+            )
```
[the remark this answers is not in the document]
Comment (joozhan) [reply python/sglang/srt/utils.py]: done

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 8]
```diff
@@ -489,8 +493,12 @@ def set_kv_buffer(
                 self.v_buffer[layer_id - self.start_layer][loc] = cache_v
             current_stream.wait_stream(self.alt_stream)
         else:
-            self.k_buffer[layer_id - self.start_layer][loc] = cache_k
-            self.v_buffer[layer_id - self.start_layer][loc] = cache_v
+            if _is_npu and loc.size() > 0 and loc.ndim == 1:
+                torch_npu.scatter_update_(self.k_buffer[layer_id - self.start_layer].unsqueeze(0), loc[0], cache_k.unsqueeze(0))
+                torch_npu.scatter_update_(self.v_buffer[layer_id - self.start_layer].unsqueeze(0), loc[0], cache_v.unsqueeze(0))
```
[the remark this answers is not in the document]
Comment (joozhan) [reply python/sglang/srt/mem_cache/memory_pool.py]: done
